prov/gni/src/gnix_mr.c

```diff
@@ -38,11 +38,210 @@
+{
+	int64_t m = 1UL << (len - 1);
+	int64_t r = (val ^ m) - m;
+
+	return r;
+}
+
+static inline gni_return_t __gnix_mr_free(gnix_mr_t *mr)
+{
+	gni_return_t ret;
+
+	ret = GNI_MemDeregister(mr->nic->gni_nic_hndl, &mr->md.mem_hndl);
+	if (ret == GNI_RC_SUCCESS) {
+		atomic_dec(&mr->md.domain->ref_cnt);
+		atomic_dec(&mr->nic->ref_cnt);
+		free(mr);
+	} else {
+		GNIX_WARN(FI_LOG_MR, "failed to deregister memory"
+				" region, mr=%p ret=%i", mr, ret);
+	}
+
+	return ret;
+}
+
+__attribute__((unused))
+static inline void __gnix_mr_put(gnix_mr_t *mr)
+{
+	if (!atomic_dec(&mr->ref_cnt)) {
+		__gnix_mr_free(mr);
+	}
+}
+
+__attribute__((unused))
+static inline void __gnix_mr_get(gnix_mr_t *mr)
+{
+	atomic_inc(&mr->ref_cnt);
+}
+
+__attribute__((unused))
+static void gnix_convert_key_to_mhdl(gnix_mr_key_t *key, gni_mem_handle_t *mhdl)
+{
+	/*
+
+	uint64_t va = (uint64_t) sign_extend(key->va << PAGE_SHIFT,
+			GNIX_MR_KEY_BITS);
+
+	GNI_MEMHNDL_INIT((*mhdl));
+	if (key->format)
+		GNI_MEMHNDL_SET_FLAGS(mdhl, GNI_MEMHNDL_FLAG_NEW_FRMT);
+	GNI_MEMHNDL_SET_VA((*mhdl), va);
+	GNI_MEMHNDL_SET_MDH((*mhdl), key->mdd);
+	GNI_MEMHNDL_SET_NPAGES((*mhdl), GNI_MEMHNDL_NPGS_MASK);
+	GNI_MEMHNDL_SET_PAGESIZE((*mhdl), PAGE_SHIFT);
+	GNI_MEMHNDL_SET_CRC(*mhdl);
+	*/
+}
+
+static void gnix_convert_mhdl_to_key(gni_mem_handle_t *mhdl, gnix_mr_key_t *key)
+{
+	//key->va = GNI_MEMHNDL_GET_VA((*mhdl));
+	//key->format = 0;
+	//key->mdd = GNI_MEMHNDL_GET_MDH((*mhdl));
+}
 
 int gnix_mr_reg(struct fid *fid, const void *buf, size_t len,
 		uint64_t access, uint64_t offset, uint64_t requested_key,
 		uint64_t flags, struct fid_mr **mr_o, void *context)
 {
-	/* TODO: implement this puppy */
-	return -FI_ENOSYS;
+	gnix_mr_t *mr;
+	int fi_gnix_access = 0;
+	struct gnix_fid_domain *domain;
+	struct gnix_nic *nic;
+	gni_cq_handle_t cq_hndl = NULL;
+	gni_return_t grc = GNI_RC_INVALID_PARAM;
+
+	if (flags)
+		return -FI_EBADFLAGS;
+
+	/* The offset parameter is reserved for future use and must be 0. */
+	if (offset || !buf || !mr_o || !access ||
+			(access & ~(FI_READ | FI_WRITE | FI_RECV | FI_SEND |
+						FI_REMOTE_READ |
+						FI_REMOTE_WRITE)) ||
+			(fid->fclass != FI_CLASS_DOMAIN))
+		return -FI_EINVAL;
+
+	/* requested key is not permitted at this point */
+	if (requested_key)
+		return -FI_EKEYREJECTED;
+
+	domain = container_of(fid, struct gnix_fid_domain, domain_fid.fid);
+
+	mr = calloc(1, sizeof(*mr));
+	if (!mr)
+		return -FI_ENOMEM;
+
+	/* If network would be able to write to this buffer, use read-write */
+	if (access & (FI_RECV | FI_READ | FI_REMOTE_WRITE))
+		fi_gnix_access |= GNI_MEM_READWRITE;
+	else
+		fi_gnix_access |= GNI_MEM_READ_ONLY;
+
+	list_for_each(&domain->nic_list, nic, list)
+	{
+
+		/* TODO: when we determine the progress model,
+		 *         set the cq_hndl pointer
+		 */
+		grc = GNI_MemRegister(nic->gni_nic_hndl, (uintptr_t) buf, len,
+					cq_hndl, fi_gnix_access,
+					-1, &mr->md.mem_hndl);
+		if (grc == GNI_RC_SUCCESS)
+			break;
```

Comment:
I can just replace that with a NULL value instead.
-- Jim




On 5/12/15, 12:57 PM, "Howard Pritchard" <notifications@github.com> wrote:

>In 
>prov/gni/src/gnix_mr.c
><https://github.com/ofi-cray/libfabric-cray/pull/155#discussion_r30161497>
>:
>> +	if (access & (FI_RECV | FI_READ | FI_REMOTE_WRITE))
>> +		fi_gnix_access |= GNI_MEM_READWRITE;
>> +	else
>> +		fi_gnix_access |= GNI_MEM_READ_ONLY;
>> +
>> +	list_for_each(&domain->nic_list, nic, list)
>> +	{
>> +
>> +		/* TODO: when we determine the progress model,
>> +		 *         set the cq_hndl pointer
>> +		 */
>> +		grc = GNI_MemRegister(nic->gni_nic_hndl, (uintptr_t) buf, len,
>> +					cq_hndl, fi_gnix_access,
>> +					-1, &mr->md.mem_hndl);
>> +		if (grc == GNI_RC_SUCCESS)
>> +			break;
>i'm pretty sure we're not going to need to use a cq_hndl for memory
>registered using the fi_mr_reg interface.
>—
>Reply to this email directly or
>view it on GitHub 
><https://github.com/ofi-cray/libfabric-cray/pull/155/files#r30161497>.
>
>